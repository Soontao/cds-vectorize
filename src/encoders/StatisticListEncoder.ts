import { variance } from "../utils/variance.js";
import Encoder from "./Encoder.js";
import { DecodeNotSupportedError } from "./Errors.js";
import ObjectMetadata from "./Metadata.js";
import ObjectEncoder, { sortMetaAndFillEncoders } from "./ObjectEncoder.js";
import { Vector } from "./type.js";
import { isNull, isNullVector, nullVector } from "./util.js";

type StatisticFunc = (values: Array<number>) => number;

const diff: StatisticFunc = (values) => {
  const max = Math.max(...values);
  const min = Math.min(...values);
  return max - min;
};

const sum: StatisticFunc = (values) => values.reduce((pre, cur) => pre + cur, 0);

const statisticFuncs: { [col: string]: StatisticFunc } = {
  max: (values) => Math.max(...values),
  min: (values) => Math.min(...values),
  avg: (values) => sum(values) / values.length,
  diff,
  variance,
  sum,
};

const statisticFuncNames = Object.keys(statisticFuncs);
const statisticColsNum = statisticFuncNames.length;

/**
 * @human
 */
export class StatisticListEncoder<T> implements Encoder<Array<T>> {
  #meta: ObjectMetadata;

  #encoder: ObjectEncoder<T>;

  #features!: Array<string>;

  constructor(meta: ObjectMetadata) {
    this.#meta = sortMetaAndFillEncoders(meta);
    this.#encoder = new ObjectEncoder(this.#meta);
  }

  features(name: string = "root"): string[] {
    if (this.#features == undefined) {
      const statisticFeatures = new Array(this.#encoder.length * statisticColsNum);
      for (const [funcIdx, funcName] of statisticFuncNames.entries()) {
        for (const [idx, objFeat] of this.#encoder.features(name).entries()) {
          const aggValueLocate = this.#encoder.length * funcIdx + idx;
          statisticFeatures[aggValueLocate] = `${objFeat}_${funcName}`;
        }
      }
      this.#features = statisticFeatures;
    }

    return this.#features;
  }

  encode(value: T[]): Vector {
    if (isNull(value) || value?.length == 0) {
      return nullVector(this.length);
    }
    const vectors = value.map((v) => this.#encoder.encode(v));
    const statisticVector = new Array(this.#encoder.length * statisticColsNum);
    for (const [funcIdx, name] of statisticFuncNames.entries()) {
      const statisticFunc = statisticFuncs[name];
      for (let idx = 0; idx < this.#encoder.length; idx++) {
        const aggValueLocate = this.#encoder.length * funcIdx + idx;
        const aggValue = statisticFunc(vectors.map((v) => v[idx]));
        statisticVector[aggValueLocate] = aggValue;
      }
    }
    return statisticVector;
  }

  decode(_vec: Vector): T[] {
    if (isNullVector(_vec)) {
      return null as any;
    }
    throw new DecodeNotSupportedError();
  }

  get length(): number {
    return this.#encoder.length! * statisticColsNum;
  }
}

export default StatisticListEncoder;
