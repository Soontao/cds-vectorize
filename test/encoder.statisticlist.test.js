import { StatisticListEncoder } from "../src/encoders/StatisticListEncoder.js";

describe("StatisticListEncoder Test Suite", () => {

  const meta = {
    properties: [
      { name: "age", type: "numeric" },
      { name: "height", type: "numeric" },
    ],
  };

  const encoder = new StatisticListEncoder(meta);

  const testData = [
    { age: 25, height: 180 },
    { age: 30, height: 175 },
    { age: 35, height: 185 },
  ];

  it("should encode correctly", () => {
    const encoded = encoder.encode(testData);
    const features = encoder.features("list_a")
    // Check if the encoded vector has the correct length
    expect(encoded.length).toBe(12);
    expect(features).toMatchSnapshot()
  });

  it("should return features correctly", () => {
    const features = encoder.features("age");

    expect(features).toBeTruthy();
    expect(features.length).toBe(12);
  });


});

// >> AI Test Cases

// Mock metadata for product information
const productMeta = {
  properties: [
    { name: "id", type: "numeric" },
    { name: "price", type: "numeric" },
    { name: "quantity", type: "numeric" },
  ],
};

// Mock data for product information
const products = [
  { id: 1, price: 1200, quantity: 5 },
  { id: 2, price: 800, quantity: 10 },
  { id: 3, price: 100, quantity: 20 },
];

// Mock metadata for statistics on product prices
const statisticsMeta = {
  properties: [
    { name: "price_max", type: "numeric" },
    { name: "price_min", type: "numeric" },
    { name: "price_avg", type: "numeric" },
  ],
};

describe("StatisticListEncoder", () => {
  let statisticListEncoder = new StatisticListEncoder(productMeta);

  test("features method should return an array of strings", () => {
    const features = statisticListEncoder.features();
    expect(Array.isArray(features)).toBe(true);
    expect(features.length).toBeGreaterThan(0);
    // Add more specific assertions based on your use case
  });

  test("encode method should return a vector with the correct length", () => {
    const encodedVector = statisticListEncoder.encode(products);
    expect(encodedVector).toBeInstanceOf(Array);
    expect(encodedVector.length).toBe(statisticListEncoder.length);
    // Add more specific assertions based on your use case
  });

  test("decode method should throw an error", () => {
    expect(() => statisticListEncoder.decode([])).toThrow("Method not supported.");
    // Add more specific assertions based on your use case
  });

});

describe("Integration with Statistics Metadata", () => {
  let statisticListEncoderWithStats = new StatisticListEncoder(statisticsMeta);

  test("features method should return an array of strings for statistics metadata", () => {
    const features = statisticListEncoderWithStats.features();
    expect(Array.isArray(features)).toBe(true);
    expect(features.length).toBe(statisticListEncoderWithStats.length);
    // Add more specific assertions based on your use case
  });

  test("encode method should return a vector with the correct length for statistics metadata", () => {
    const encodedVector = statisticListEncoderWithStats.encode(products);
    expect(encodedVector).toBeInstanceOf(Array);
    expect(encodedVector.length).toBe(statisticListEncoderWithStats.length);
  });

});
