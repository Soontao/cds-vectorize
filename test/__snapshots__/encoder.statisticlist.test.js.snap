// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticListEncoder Test Suite should encode correctly 1`] = `
[
  "list_a_age_max",
  "list_a_height_max",
  "list_a_age_min",
  "list_a_height_min",
  "list_a_age_avg",
  "list_a_height_avg",
  "list_a_age_diff",
  "list_a_height_diff",
  "list_a_age_variance",
  "list_a_height_variance",
  "list_a_age_sum",
  "list_a_height_sum",
]
`;

exports[`StatisticListEncoder Test Suite should support empty array 1`] = `
[
  -65536,
  -65536,
  -65536,
  -65536,
  -65536,
  -65536,
  -65536,
  -65536,
  -65536,
  -65536,
  -65536,
  -65536,
]
`;
